---
title: "528Assignment4_PSSM"
author: "Meiheng Liang"
date: "2024-04-10"
output: html_document
---

```{r setup, include=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
if (!requireNamespace("rentrez", quietly = TRUE))
    install.packages("rentrez")
BiocManager::install("Biostrings")
library(Biostrings)

```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r prep data}
# Reading a comma-separated .txt file
data <- read.csv("C:/Users/Chris/git-test/argR-counts-matrix.txt", header = FALSE, sep = "\t", fill = TRUE, strip.white = TRUE)
rownames(data) <- data[,1]  # Set the first column as row names
rownames(data) <- toupper(rownames(data))
matrix <- data[,-c(1,2)]

```

## Including Plots
```{r frequency matrix }
matrix.freq <- (matrix ) / apply(matrix, 2, sum)
matrix.freq 
```


```{r pseudo frequency matrix}
k <- 1 
Pseudo.freq <- (matrix + k) / apply(matrix + k, 2, sum)
Pseudo.freq
weights <- log(Pseudo.freq / 0.25)
weights
```

```{r data prep: top 30 gene}
inputFilePath <- ("C:/Users/Chris/git-test/E_coli_K12_MG1655_.400_50")
outputFilePath <- ("C:/Users/Chris/git-test/E_coli_K12.fasta")
# Read the raw data
rawfile <- readLines(inputFilePath)
# Open a connection to the output file
outputConnection <- file(outputFilePath, "w")
# Process each line
for(line in rawfile) {
  # Split the line into ID and sequence parts
  parts <- strsplit(line, " \\\\ ")[[1]]  # Note the double escape for backslash
  
  # Write the formatted entry to the output file
  cat(sprintf(">%s\n%s\n", parts[1], parts[2]), file = outputConnection)
}

# Close the connection
close(outputConnection)

E.coli <- readDNAStringSet ("C:/Users/Chris/git-test/E_coli_K12.fasta")
head(E.coli)
```


```{r screening: top 30 gene}
scoreSequenceWithPSWM <- function(sequence, pswm) {
  sequence <- as.character(sequence)
  maxScore <- -Inf  # Start with a very low number that any real score will exceed
  maxPosition <- 0
  seqLength <- nchar(sequence)
  pswmWidth <- ncol(pswm)
  if (seqLength < pswmWidth) {
    return(list(maxScore = NA, maxPosition = NA))  # If sequence is too short, return NA
  }

  for (pos in 1:(seqLength - pswmWidth + 1)) {
    subseq <- substring(sequence, pos, pos + pswmWidth - 1)
    tempScore <- 0
    for (i in 1:pswmWidth) {
      nucleotide <- substr(subseq, i, i) # (sequence, start, end)
      if (!is.na(pswm[nucleotide, i])) {
        tempScore <- tempScore + pswm[nucleotide, i]
      } else {
        tempScore <- NA  # Handle cases where nucleotide is not in pswm
        break
      }
    }
    # Only update if tempScore is not NA and greater than maxScore
    if (!is.na(tempScore) && tempScore > maxScore) {
      maxScore <- tempScore
      maxPosition <- pos
    }
  }

  if (maxScore == -Inf) {  # No valid scores found, return NA
    maxScore <- NA
    maxPosition <- NA
  }

  return(list(maxScore = maxScore, maxPosition = maxPosition))
}



GeneList <- lapply(E.coli, scoreSequenceWithPSWM, pswm=weights)
GeneList
scoresList <- do.call(rbind, lapply(GeneList, function(x) {
    data.frame(maxScore = x$maxScore, sequenceposition = x$maxPosition)
}))
GeneRank <- scoresList[order(scoresList$maxScore, decreasing = TRUE), , drop = FALSE]
top30 <- GeneRank[1:30, , drop = FALSE]
top30


```

